Support for Shelly 1 temp. addon

CurlyMoo opened the pull request.

TD-er committed:
[Cleanup] Wemos motor shield, reduce resources used.

TD-er committed:
[docs] Minor Sphinx warnings fixed.

TD-er committed:
Make sure notifications (Email/Buzzer) are included in 1M normal build

TD-er committed:
[IRTX] Use parse argument functions for IR TX command arguments

As mentioned here: https://github.com/letscontrolit/ESPEasy/issues/2724#issuecomment-735778417

TD-er committed:
[NeoClock] Fix building P070 NeoPixelClock

As reported on [the forum](https://www.letscontrolit.com/forum/viewtopic.php?f=7&t=8126&p=48439#p48439)

TD-er committed:
[PVS-Studio] Allow testing IR builds for coding errors

TD-er committed:
[IRTX] Convert more commands to use parse string functions

CurlyMoo committed:
Support for Shelly 1 temp. addon

CurlyMoo commented:
The code for the dallas sensor was adapted to work with the Shelly 1 temperature addon. This is of course not a permanent solution, because it breaks the regular dallas 1-wire functionality. However, it does show what's needed to get it to work on the Shelly.

I used the code from Tasmota as a reference. They fixed it by adapting the abstract dallas code base to work either on 1 wire (default) or two (shelly). I'm not familiar enough with the ESPEasy libraries to properly implement an option for either 1 or 2 wire configurations, but i do agree that would be the way to go.

TD-er commented:
How is the Dallas connected via an RX and TX?
It is a 1-wire  protocol.

Or is the Vdd of the Dallas connected to a GPIO pin?

TD-er commented on src/src/Helpers/Dallas1WireHelper.cpp:
```diff
@@ -619,10 +625,9 @@ uint8_t Dallas_reset(int8_t gpio_pin)
-    pinMode(gpio_pin, OUTPUT);
+    digitalWrite(gpio_pin_tx, LOW);
     delayMicroseconds(480);
-    pinMode(gpio_pin, INPUT); // Float
```
Having 2 pins defined and adding a selector to "copy" the RX pin to TX pin on initialization when using a single pin can still work with the 1 pin configuration.
However removing this pin mode statement to set the pin to floating will for sure break the 1-pin configuration.

TD-er commented on src/src/Helpers/Dallas1WireHelper.cpp:
```diff
@@ -844,40 +849,40 @@ uint8_t Dallas_read(int8_t gpio_pin)
     // Wait for 6 usec
   }
   unsigned long startwait = micros();
-  pinMode(gpio_pin, INPUT); // let pin float, pull up will raise
```
The same here.
The pin MUST be floating or else the 1-pin config will no longer work.

CurlyMoo replied:
>  The pin MUST be floating or else the 1-pin config will no longer work.

Of course, that's what i already said in my initial comment:

> This is of course not a permanent solution, because it breaks the regular dallas 1-wire functionality. 

Tasmota fixes this by allowing the DS18B20 library to either run in single or dual mode:
https://github.com/arendst/Tasmota/blob/development/tasmota/xsns_05_ds18x20.ino#L81

CurlyMoo commented:
> 
> 
> How is the Dallas connected via an RX and TX?
> It is a 1-wire protocol.
> 
> Or is the Vdd of the Dallas connected to a GPIO pin?

This is the best summary i can give: https://github.com/arendst/Tasmota/issues/7469

TD-er commented:
Ah now I get it, since the module uses an isolator you need to split RX and TX.
This can be added indeed, like you did.
But we also need to add more logic to support both 1 and 2 GPIO pins.
As you did now, it will not work anymore on 1 pin setups, even if you set RX and TX to use the same pin.

CurlyMoo commented:
>  But we also need to add more logic to support both 1 and 2 GPIO pins.

Indeed

>  As you did now, it will not work anymore on 1 pin setups, even if you set RX and TX to use the same pin.

True, the logic can be that if both RX and TX are the same, it uses a one pin setup and when they are different a two pin setup. Or, an additional checkbox that differentiates between the two setups.

TD-er commented:
> True, the logic can be that if both RX and TX are the same, it uses a one pin setup and when they are different a two pin setup. Or, an additional checkbox that differentiates between the two setups.

Or simply consider the 2nd pin as optional and if not set (or equal) consider it as a 1-pin setup.

CurlyMoo commented:
> Or simply consider the 2nd pin as optional and if not set (or equal) consider it as a 1-pin setup.

Also a good idea.

TD-er committed:
Fix missing defines to allow "Periodical send Gratuitous ARP"

CurlyMoo commented:
My Shelly 1 is already put in production to control my domestic hot water tank. So, if someone can test any updated code for me?

I will try to incoorporate the idea's posted here today.

TD-er committed:
[Cleanup] PluginStructs/*.cpp include prepend PluginStrucs dir

And let the file end with a newline to make sure concatenating these files will not result in build errors.

TD-er committed:
[PIO] Concatenate .cpp files to prevent linking errors

TD-er committed:
[PIO] Concat *.cpp files in ControllerQueue/ and PluginStructs/

CurlyMoo commented:
@TD-er I've pushed a new commit incorporating what we discussed. The TX pin is now optional. If the TX and RX pins are the same (which will be forcefully synced if the TX pin isn't set) it will use single mode. If the TX and RX pins are different it will use the dual mode.

TD-er committed:
[PIO] Add Commands/*.cpp to the files to be concatenated in single .cpp

TD-er commented:
Can you please add commits to the branch instead of force pushing it?
Now I don't see the changes you made and have to look at all the code again (and again...)

TD-er committed:
Fix build error in Linux build environment

Define clashing with variable name.

CurlyMoo commented:
> Can you please add commits to the branch instead of force pushing it?

Nearly all projects i contributed to (included my own) required a to a single commit rebased PR.

The changes are very minimal. The `pinModes` are back again based on `gpio_pin_rx == gpio_pin_tx` and the `tx` pin is now optional.

CurlyMoo commented on src/_P004_Dallas.ino:
```diff
@@ -114,17 +114,22 @@ boolean Plugin_004(byte function, struct EventStruct *event, String& string)
     {
-      event->String1 = formatGpioName_bidirectional(F("1-Wire"));
+      event->String1 = formatGpioName_RX(false);
+      event->String2 = formatGpioName_TX(true);
```
This is new

CurlyMoo commented on src/_P004_Dallas.ino:
```diff
@@ -114,17 +114,22 @@ boolean Plugin_004(byte function, struct EventStruct *event, String& string)
-      int8_t Plugin_004_DallasPin = CONFIG_PIN1;
+      int8_t Plugin_004_DallasPin_RX = CONFIG_PIN1;
+      int8_t Plugin_004_DallasPin_TX = CONFIG_PIN2;
+      if(Plugin_004_DallasPin_TX == -1) {
```
This is new

CurlyMoo commented on src/_P004_Dallas.ino:
```diff
@@ -176,11 +181,15 @@ boolean Plugin_004(byte function, struct EventStruct *event, String& string)
-      int8_t Plugin_004_DallasPin = CONFIG_PIN1;
+      int8_t Plugin_004_DallasPin_RX = CONFIG_PIN1;
+      int8_t Plugin_004_DallasPin_TX = CONFIG_PIN2;
+      if(Plugin_004_DallasPin_TX == -1) {
```
This is new

CurlyMoo commented on src/_P080_DallasIButton.ino:
```diff
@@ -65,7 +65,7 @@ boolean Plugin_080(byte function, struct EventStruct *event, String& string)
 
       if (Plugin_080_DallasPin != -1) {
-        Dallas_addr_selector_webform_load(event->TaskIndex, Plugin_080_DallasPin);
+        Dallas_addr_selector_webform_load(event->TaskIndex, Plugin_080_DallasPin, Plugin_080_DallasPin);
```
All changes in this file are new

CurlyMoo commented on src/src/Helpers/Dallas1WireHelper.cpp:
```diff
@@ -122,14 +122,23 @@ void Dallas_addr_selector_webform_load(taskIndex_t TaskIndex, int8_t gpio_pin, u
+  // input and output pins, and therefor
+  // doesn't switch between input and output
+  // when running.
+  if(gpio_pin_rx != gpio_pin_tx) {
```
This is new

CurlyMoo commented on src/src/Helpers/Dallas1WireHelper.cpp:
```diff
@@ -592,14 +604,18 @@ bool Dallas_setResolution(const uint8_t ROM[8], byte res, int8_t gpio_pin)
     #if defined(ESP32)
   ESP32noInterrupts();
     #endif // if defined(ESP32)
-  pinMode(gpio_pin, INPUT);
```
This is new

CurlyMoo commented on src/src/Helpers/Dallas1WireHelper.cpp:
```diff
@@ -619,10 +635,16 @@ uint8_t Dallas_reset(int8_t gpio_pin)
-    digitalWrite(gpio_pin, LOW);
-    pinMode(gpio_pin, OUTPUT);
+    digitalWrite(gpio_pin_tx, LOW);
+    if(gpio_pin_rx == gpio_pin_tx) {
```
This is new

CurlyMoo commented on src/src/Helpers/Dallas1WireHelper.cpp:
```diff
@@ -619,10 +635,16 @@ uint8_t Dallas_reset(int8_t gpio_pin)
+    }
     delayMicroseconds(480);
-    pinMode(gpio_pin, INPUT); // Float
+    if(gpio_pin_rx == gpio_pin_tx) {
```
This is new

CurlyMoo commented on src/src/Helpers/Dallas1WireHelper.cpp:
```diff
@@ -844,40 +866,47 @@ uint8_t Dallas_read(int8_t gpio_pin)
     #endif // if defined(ESP32)
-  digitalWrite(gpio_pin, LOW);
-  pinMode(gpio_pin, OUTPUT);
+  digitalWrite(gpio_pin_tx, LOW);
```
This is new

CurlyMoo commented on src/src/Helpers/Dallas1WireHelper.cpp:
```diff
@@ -844,40 +866,47 @@ uint8_t Dallas_read(int8_t gpio_pin)
   }
   unsigned long startwait = micros();
-  pinMode(gpio_pin, INPUT); // let pin float, pull up will raise
+  if(gpio_pin_rx == gpio_pin_tx) {
```
This is new

CurlyMoo commented on src/src/Helpers/Dallas1WireHelper.cpp:
```diff
@@ -911,14 +940,16 @@ void Dallas_write_bit(uint8_t v, int8_t gpio_pin)
-  digitalWrite(gpio_pin, LOW);
-  pinMode(gpio_pin, OUTPUT);
+  digitalWrite(gpio_pin_tx, LOW);
+  if(gpio_pin_rx == gpio_pin_tx) {
```
This is new

CurlyMoo commented:
As you can see, I added a comment to all lines that are new since my initial commit.

tonhuisman commented:
>
> Nearly all projects i contributed to (included my own) required a to a single commit rebased PR.

That sounds very un-git like, and counter-productive. 🤔 
Sure you can choose to rebase before merging a PR, but during the creation of the branch, multiple commits can be made, even after creating the PR. Especially when review is desired/required the only (easy) way of sharing your work on github is to create a PR. Otherwise you would have to constantly invite project maintainers to your personal clone with your branch to do any review. (And sometimes even create a clone of the clone)

TD-er commented:
In essence a PR is a request to merge a branch.
But for merging it doesn't really matter if this is 1 commit or N commits.
However if you have a review process like we have now, it is very useful to have several commits.
This allows to see the diffs per commit and also for the entire PR.
If a maintainer only wants to see the commits squashed into one when merging, you can simply check the checkbox to squash the commits at merge.

By appending to the existing commit and then force push'ing it, you keep having a single commit. (the same when you rebase the branch and force push it instead of pull first)

But if you create a commit for every change and push it, you simply include the commit to the pending PR.

CurlyMoo commented:
> That sounds very un-git like, and counter-productive

> In essence a PR is a request to merge a branch.

I very well understand how git works. If you want me to implement another gitflow i of course will.

> even after creating the PR

That feature is quite new. Therefor, a lot of project i contributed to still have the old gitflow in place. A PR that should be kept rebased to a single commit. Indeed, with the new rebasing option of the PR, that necessity is gone.

----

I know how git works, and due to the lack of contribution guide i just did what i'm familiar with ever since i use git. Let us focus on the actual content of the PR.

TD-er commented:
> I know how git works, and due to the lack of contribution guide i just did what i'm familiar with ever since i use git. Let us focus on the actual content of the PR.

Sure and I looked over it and so far it looks OK
I still have to test it on a setup with 1 wire connection to make sure nothing is broken here.

TD-er committed:
[PIO] Separate appended .cpp files with newline

And add #ifdef USES_Cxxx checks for queue element code.

TD-er committed:
[PIO] Concatenate all WebServer/*.cpp files into temp file.

And add some build output formatting

TD-er committed:
[PIO] Make sure library includes are present in .cpp 

.h files are in the ignored directory, so the PIO library finder may not mark the library as being used resulting in a failed build.

CurlyMoo commented:
I've tested the code myself in single mode. It works, but not perfectly. Somehow i need to set both pins in order to get it to work the first time. As soon it started reading the first values the optional pin can remain blank. When setting just one pin i also get the message `DS   : Cannot set resolution` although i don't see how just setting one pin is currently different from setting both to the same value.

TD-er commented:
Setting the resolution is performing a write to the chip, which does take a minimum amount of time before you can send a new command.
How is the sensor connected in your test setup?
A separate power line to the Vdd pin of the sensor?
What voltage are you using? The Dallas sensor has a minimum allowed voltage of 3.0V if I'm not mistaken.

CurlyMoo commented:
> How is the sensor connected in your test setup?

In my simple test directly.

>  A separate power line to the Vdd pin of the sensor?

Yes

> What voltage are you using? The Dallas sensor has a minimum allowed voltage of 3.0V if I'm not mistaken.

5v

TD-er commented:
Need also to add this plugin to the Shelly1 build definition.

CurlyMoo commented:
> Need also to add this plugin to the Shelly1 build definition.

This wasn't enough?
https://github.com/letscontrolit/ESPEasy/blob/596d020a56286a64a630bb3f1cfc270740c03da0/src/src/CustomBuild/define_plugin_sets.h#L452

TD-er commented:
> > Need also to add this plugin to the Shelly1 build definition.
> 
> This wasn't enough?
> https://github.com/letscontrolit/ESPEasy/blob/596d020a56286a64a630bb3f1cfc270740c03da0/src/src/CustomBuild/define_plugin_sets.h#L452

Yes it is...
It was more like a mental note for myself based on the request made in issue #3414

TD-er committed:
[Web UI] Proper encoding for HTML attribute values

All caused by this little remark: https://github.com/letscontrolit/ESPEasy/pull/3404#issuecomment-737972401

TD-er committed:
[Cleanup] Uncrustify of web page code

TD-er committed:
[Dallas] Fix P100 DS2423 counter plugin matching new changes.

TD-er committed:
[Dallas] Fix P100 DS2423 counter plugin matching new changes.

TD-er commented:
The P100 plugin was not yet updated, so I added that one too.
It is now running on my breadboard node with 6 Dallas chips using a single pin and that's working fine so it seems.

Only thing to be done now is document what the second pin may be used for.

TD-er committed:
Merge branch 'mega' into shelly-ds18b20

TD-er committed:
Merge branch 'mega' into shelly-ds18b20

CurlyMoo commented:
> Only thing to be done now is document what the second pin may be used for.

And having other users verify the Shelly 1 Addon functionality. Mine was based on the original PR code.

> Only thing to be done now is document what the second pin may be used for.

Should i do it or will you?

TD-er commented:
> > Only thing to be done now is document what the second pin may be used for.
> 
> Should i do it or will you?

Be my guest :)

CurlyMoo committed:
Added documentation

CurlyMoo committed:
Merge branch 'shelly-ds18b20' of https://github.com/CurlyMoo/ESPEasy into shelly-ds18b20

CurlyMoo committed:
Small typo in documentation

CurlyMoo committed:
Merge branch 'shelly-ds18b20' of https://github.com/CurlyMoo/ESPEasy into shelly-ds18b20

TD-er commented on docs/source/Plugin/P004_DS18b20.rst:
```diff
@@ -93,10 +93,15 @@ Sensor
+GPIO <--> Shelly Temperature add-on
+"""""""""""""""""""""""""""""""""""
+
+The Shelly temperature add-on require a two wire setup. Set TX to GPIO 3 and RX to GPIO 0.
```
One question about this.
Does that cause any conflict with Serial0?

CurlyMoo replied:
I don't know. What can i do to check?

TD-er replied:
Does it still work when "Use Serial" is enabled in tools -> Advanced?
What happens if you have the Dallas plugin still enabled and trying to connect via serial?

CurlyMoo replied:
> What happens if you have the Dallas plugin still enabled and trying to connect via serial?

I don't think that matters much because there is no sensible way you want to have both enabled at the same time on a Shelly.

> Does it still work when "Use Serial" is enabled in tools -> Advanced?

Yes it does.

TD-er replied:
Well if "Use Serial" is enabled, any data received on that port, which may somehow be interpreted as data, is handed over to the command interpreter.
So it may end up giving garbage in the logs, or at least cause extra load for the parsing.
Also I don't know what effect sending the logs to serial may have on the operations of the Dallas task.

Not saying it is useful to have them active at the same time, but at least there should be a warning in the documentation about disabling serial when using those pins for other tasks.

CurlyMoo replied:
> but at least there should be a warning in the documentation about disabling serial when using those pins for other tasks.

Isn't that the reason that those pins have a warning sign attached in the first place. I mean the warning should be there in general, not specified to this plugin. As in: `When using a warning sign GPIO, make sure you disable the accompanied feature attached to it`

TD-er replied:
Yep the warning sign is there if the pins are also actively used via "Use Serial".
But maybe some mentioning in the documentation that GPIO-3 might be used by the serial port can give people a hint where to look.

The warning sign is often just ignored by users. So there has to be some kind of trade-off between "explaining it all" and "being flooded by reported issues or forum questions".

CurlyMoo replied:
> But maybe some mentioning in the documentation that GPIO-3 might be used by the serial port can give people a hint where to look.

On the `Shelly` section of the `004` page?

TD-er replied:
Yep. For others it doesn't really apply.
But on the Shelly you mention very specific pins to use, so users are bound to run into issues as the "Use Serial" is enabled by default.

CurlyMoo committed:
Small comment about potential serial port conflicts

CurlyMoo commented:
Didn't you want additional confirmation for the Shelly functionality?

TD-er commented:
I will not yet make a nightly build.
That's probably not due till Monday.
And for the existing users the plugin is still working.

CurlyMoo commented:
> Sure you can choose to rebase before merging a PR

> If a maintainer only wants to see the commits squashed into one when merging, you can simply check the checkbox to squash the commits at merge.

🤔 😉
